README.md
# ARIA Buttons
This script largely serves as a demonstration of a way to ensure that elements with `role=button` can be activated using <kbd>Space</kbd> and <kbd>Enter</kbd> keys.  It makes sure that if a `role=button` has an `aria-disabled=true`, then it will _not_ be keyboard focusable.  Finally, if the script detects an `aria-pressed` attribute, it will assign a function specifically to toggle the value of the attribute.


## Why do you need this?  
As long as you're setting up `role="button"` elements to work with the appropriate keyboard controls in your own scripts, then you likely don't need this.
If you are using native HTML `<button>`s, then you also likely don't need this.

In general, you likely don't need this.  Just use a `<button>` and manage your own `aria-pressed` :)


## License & Such  
This script was written by [Scott O'Hara](https://twitter.com/scottohara).

It has an [MIT license](https://github.com/scottaohara/accessible-components/blob/master/LICENSE.md).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
...

## 2.0.0 - 2021-12-11
### Changed
- Revised and simplified script. 
- Removed use of `data-*` attributes. Script instead just looks for use of `role=button` and `aria-pressed`.
- Added checks for `aria-disabled` and revise `tabindex` checks based on this attribute.

## 1.1.0 - 2018-08-26
### Changed
- Remove `href` attributes from links that have been converted to `role=button`. Doing so will ensure that a link context menu won't appear if someone right clicks on a "button."
- Change 'keytrolls' function to 'keyEvents'.
- Add `tabindex=0` to all converted elements, since the condition for `href` no longer exists.

## 1.0.0 - 2018-02-08
### Changed
- Update documentation and features. Declare script version 1.0.0.

### Added
- `aria-pressed` and `data-pressed` functionality added to script.


## 0.3.1 - 2017-03-31
### Changed
- Add `undefined` to iffe
- Revised naming of function from `a11yBUTTONS` to `ARIAbuttons`.
- Remove `widgetCount` variable.


## 0.3.0 - 2017-03-24
### Changed
- Move `var widget` to be within the `create` function.  Add `default break` to switch.


## 0.2.0 - 2016-12-17
### Fixed
- Needed a `preventdefault()` for link buttons (e.g., `a href`, to keep focus on the element when triggered).


## 0.1.0 - 2016-11-26
### Added
- Initial commit of code base.
- Progressively enhance non-button elements into a `role=button`. 
- Script adds `tabindex=0` if necessary.
- Script adds ENTER and SPACE keys to activate "button".
- Script adds  `aria-controls` attribute if `data-controls` or `href` existed.


index.js
(function ( w, doc, undefined ) {
  /**
   * ARIA Buttons
   *
   * Author: Scott O'Hara
   * Version: 2.0.0
   * License: https://github.com/scottaohara/aria_buttons/blob/master/LICENSE
   */

  const ariaButton = function ( inst, options ) {
    const el = inst;
    let isToggle;

    /**
     * Initialize the button instance
     */
    const init = function () {
      setupButton();

      if ( isToggle ) {
        el.addEventListener('click', toggleEvent);
      }
      el.addEventListener('keydown', keyEvents);
    };


    const setupButton = function () {
      if ( !el.hasAttribute('tabindex') && el.getAttribute('aria-disabled') !== 'true' ) {
        el.tabIndex = 0;
      }
      if ( el.hasAttribute('tabindex') && el.getAttribute('aria-disabled') === 'true' ) {
        el.removeAttribute('tabindex');
      }
      if ( el.hasAttribute('aria-pressed') ) isToggle = true;
      /**
       * If an element started off as a <a href...> remove the href attribute.
       * Buttons should not return a context menu for links, if right clicked.
       */
      el.removeAttribute('href');
    };


    const toggleEvent = function ( e ) {
      if ( isToggle ) {
        el.setAttribute('aria-pressed', el.getAttribute('aria-pressed') === 'true' ? 'false' : 'true');
      }
    };


    const keyEvents = function ( e ) {
      const keyCode = e.keyCode || e.which;

      switch ( keyCode ) {
        case 32:
        case 13:
          e.preventDefault();
          this.click();
          break;

        default:
          break;
      }
    }


    init.call( this );
    return this;
  }; // ariaButton()

  w.ariaButton = ariaButton;
})( window, document );


assets/js/a11y.buttons.js
;(function ( w, doc, undefined ) {
  'use strict';

  /**
   * Local object for method references
   * and define script meta-data
   */
  var ARIAbuttons = {};
  w.ARIAbuttons   = ARIAbuttons;

  ARIAbuttons.NS      = 'ARIAbuttons';
  ARIAbuttons.AUTHOR  = 'Scott O\'Hara';
  ARIAbuttons.VERION  = '1.1.0';
  ARIAbuttons.LICENSE = 'https://github.com/scottaohara/aria_buttons/blob/master/LICENSE';

  /**
   * Create Button Instances
   */
  ARIAbuttons.create = function () {
    var widget = doc.querySelectorAll('[role="button"]');
    var self;
    var i;

    // Setup all instances of role="button"
    for ( i = 0; i < widget.length; i++ ) {
      self = widget[i];

      /**
       * Buttons need to be focusable by keyboard users.
       * If no tabIndex was set, set one.
       */
      if ( !self.hasAttribute('tabindex') ) {
        self.tabIndex = 0;
      }

      /**
       * If a "button" doesn't have an aria-controls attribute,
       * do additional checks to see if it *should* have one.
       */
      if ( !self.hasAttribute('aria-controls') ) {
        var ac = 'aria-controls';

        /**
         * Check for the presence of an href (if the element was a link),
         * or check for a data-controls attribute. If one of these exist,
         * get the current value and use it as the IDREF for the
         * aria-controls attribute.
         */
        if ( self.hasAttribute('data-controls') ) {
          self.setAttribute(ac, self.getAttribute('data-controls'));
        }
        else if ( self.hasAttribute('href') ) {
          self.setAttribute(ac, self.getAttribute('href').split('#')[1]);
        }

        // clean up DOM
        self.removeAttribute('data-controls');
      } // if

      /**
       * If an element started off as a <a href...> remove the href attribute.
       * Buttons should not return a context menu for links, if right clicked.
       */
      self.removeAttribute('href');

      /**
       * Check to see if this is meant to be a toggle button.
       *
       * If the element has an aria-pressed already, move on. If it doesn't,
       * check for the presence of a data-pressed attribute. If one exists,
       * add an aria-pressed with the appropriate value.
       */
      if ( !self.hasAttribute('aria-pressed') ) {
        var ap = 'aria-pressed';

        if ( self.hasAttribute('data-pressed') && self.getAttribute('data-pressed') === 'true' ) {
          self.setAttribute(ap, 'true');
        }
        else if ( self.hasAttribute('data-pressed') ) {
          self.setAttribute(ap, 'false');
        }

        // clean up DOM
        self.removeAttribute('data-pressed');
      } // if

      /**
       * Run event listeners for each instance.
       */
      self.addEventListener('keypress', ARIAbuttons.keyEvents, false);
      self.addEventListener('click', ARIAbuttons.ariaPressed, false);
    } // for(widget.length)
  }; // ARIAbuttons.create()


  /**
   * Keyboard Controls for the 'Buttons'
   */
  ARIAbuttons.keyEvents = function ( e ) {
    var keyCode = e.keyCode || e.which;

    switch ( keyCode ) {
      // enter or space
      case 32:
      case 13:
        e.preventDefault();
        e.target.click();
        break;

      default:
        break;
    }
  }; // ARIAbuttons.keyEvents();

  /**
   * Toggle the value of aria-pressed.
   */
  ARIAbuttons.ariaPressed = function ( e ) {
    e.preventDefault();

    if ( this.getAttribute('aria-pressed') === 'true' ) {
      this.setAttribute('aria-pressed', 'false');
    }
    else {
      this.setAttribute('aria-pressed', 'true');
    }
  } // ARIAbuttons.ariaPressed()


  /**
   * Initialize Buttons Functions.
   * If expanding this script, place any other
   * initialize functions within here.
   */
  ARIAbuttons.init = function () {
    ARIAbuttons.create();
  };

  // go go JavaScript
  ARIAbuttons.init();

})( window, document );
